tests/unit/ip_router_interface/v0/test_ip_router.py

```diff
@@ -0,0 +1,533 @@
+# Copyright 2023 Ubuntu
+# See LICENSE file for licensing details.
+#
+# Learn more about testing at: https://juju.is/docs/sdk/testing
+import json
+import logging
+import os
+import shutil
+import unittest
+from itertools import chain
+
+from ops.testing import Harness
+
+from tests.provider_charm.src.charm import SimpleIPRouteProviderCharm
+from tests.requirer_charm.src.charm import SimpleIPRouteRequirerCharm
+
+logger = logging.getLogger(__name__)
+
+LIB_DIR = "lib/charms/ip_router_interface/v0"
+LIB_NAME = "ip_router_interface.py"
+REQUIRER_CHARM_DIR = "tests/requirer_charm"
+PROVIDER_CHARM_DIR = "tests/provider_charm"
+IP_ROUTER_PROVIDER_APP_NAME = "ip-router-provider"
+IP_ROUTER_REQUIRER_APP_NAME = "ip-router-requirer"
+
+
+def copy_lib_content() -> None:
+    os.makedirs(f"{REQUIRER_CHARM_DIR}/{LIB_DIR}", exist_ok=True)
+    os.makedirs(f"{PROVIDER_CHARM_DIR}/{LIB_DIR}", exist_ok=True)
+    shutil.copyfile(src=f"{LIB_DIR}/{LIB_NAME}", dst=f"{REQUIRER_CHARM_DIR}/{LIB_DIR}/{LIB_NAME}")
+    shutil.copyfile(src=f"{LIB_DIR}/{LIB_NAME}", dst=f"{PROVIDER_CHARM_DIR}/{LIB_DIR}/{LIB_NAME}")
+
+
+class TestProvider(unittest.TestCase):
+    def _setup(self):
```

Comment:
Why are you not using the standard `def setUp(self)`?

Reply:
Did not know about it :D. Fixed.